### ISSUE TITLE
TCR: Option to `--amend` previous commit (or allow configuration of the commit command)

### ISSUE THREAD

**Issue by michelchenrich** [dkandalov/limited-wip#27]
Hello  @dkandalov,

First of all, I'd like to mention that this is an amazing plug-in overall. Thank you very much for making it available for everyone! 😄

Only thing I'm wondering is if it is possible customize how the commits are added with a bit more flexibility.
For example, I'd like an option for amending the previous commit added via TCR between each push, to avoid creating a huge list of relatively meaningless commits in sequence. For instance, let's say that to refactor something a bit bigger, e.g. switch-case into polymorphic dispatch, takes about 10 to 15 TCR cycles. Either we leave all commits in sequence with the same name, or we squash those commits into one which describes the entire refactoring. In such cases it would be better to write the first commit message, and have the following TCR cycles amending it. By "first commit", I mean the first change after being in sync with `remote`, so a `git push` would be the trigger for resetting the amend behavior, asking for a new commit message in the next TCR cycle, then continuing with the amends.
 
I realize this may be a very personal style, so in that case, I'd just like to suggest way to "tweak" the git commands that are sent on each passing test run, to make this logic happen. I believe that something like a "shell script to be executed on test pass" could do the trick.

**Comment by dkandalov** [comment]
Hello Michel,

Thank you, I'm glad you like the plugin 😊

Yes, I've been thinking about something like that for a while. As I understand, what you suggest will be another option in the TCR settings "On passed test: amend commit" for which an example workflow might be:
 - open IDE
 - write some code, tests pass
 - plugin asks for commit message and commits (because IDE was just opened)
 - write some code, tests pass
 - plugin amends last commit
 - push changes
 - write some code, tests pass
 - plugin asks for commit message and commits (because there was push before last commit)

Configuring "shell script to be executed on test pass" is also an interesting idea.

**Comment by michelchenrich** [comment]
Yes, the workflow matches almost perfectly, the only thing that maybe helps simplifying the implementation would be to always check git itself for the state of the last commit, meaining that the IDE/Plugin does not need to keep track of state changes like opening/closing the IDE, performing a push, etc.

For example, when working from VIM/terminal, I have a shell script I for TCR which includes the following part after the tests pass:
```
git add .;
if [ -z $(git rev-parse HEAD | xargs git branch -r --contains) ]; then
   git commit --amend --no-edit;
else
  git commit;
fi
```
It runs `git rev-parse HEAD` to get the last commit hash in the current branch, then runs `git branch -r --contains <commit-hash>` to check if that commit is included in any of the remote branches (this runs against the local copies of the remote branches, so it doesn't go through the network). The result is a list of remote branches which contain the commit hash, which is then checked if it is empty:
- If the list is empty, it means we have a commit which was never pushed, and so we amend it.
- If the list returns anything else (except error), it means we pushed everything and should prompt for the commit message. In this case I simply use `git commit` because this runs from terminal and opens up VIM to edit the commit message. From inside IntelliJ this could be changed to open the integrated git UI prompt (this part I have no experience with), but in any case, that command should open whatever editor is configured via `git config --global core.editor <option>`.

By the way, this is one example of a script which could simply be executed on tests passing.

**Comment by dkandalov** [comment]
I like the idea of checking git for the state. Here are couple potential problems I could think of:
 - Currently, the plugin doesn't depend on any particular VCS and in theory should work with mercurial, svn or any other VCS supported by IDE (I haven't tried it though).  So, ideally, the solution shouldn't be git-specific. There is probably an option to implement this logic using API for distributed VCSs (git and mercurial).  Another option might be to have the logic you suggested for git and something else for other VCSs which probably are almost never used with the plugin anyway.
 - I was wondering if the performance of checking commits on multiple branches is good enough when there are lots of branches.  For example, IntelliJ used to have huge amount of branches (not anymore). I'm not really concerned about it though.
 - What if I want to start working on a new feature with TCR but I don't have internet connection to push?  There should be some other way to tell plugin to not amend next time tests are passing.
 Maybe this could be done on invoking `Commit` action which is currently disabled in TCR mode 🤔

As an idea maybe checking if the list of tests which were run are different from the last time, can be a good indicator of a new feature. I think it'll work most of the time for me.

After thinking for a bit about executing scripts on passed/failing tests, it seems to be a more generic idea than just TCR customisation. Maybe IntelliJ plugin which could run shell scripts on other IDE "events" like VCS update or failed compilation.

And why stop at receiving data from IDE? The plugin could also specify an http API to invoke IDE actions from shell scripts.This could be a way to extend IDE at runtime with external scripts.
With such a plugin it should be possible to implement your own custom TCR with couple callbacks to scripts. (Maybe I could make this part of https://github.com/dkandalov/live-plugin one day 🤔)

**Comment by dkandalov** [comment]
FYI I implemented `commit amend` (via IntelliJ API) if the last commit only exists on the current branch (as you suggested) **and** executed test name is the same as the last executed test (this is to be able to start new commit offline without pushing).

I'll try it for some time and if it looks ok, will release the plugin.

Not sure if you can access it, but here is the release draft with zip file
https://github.com/dkandalov/limited-wip/releases/tag/untagged-9014f0dcba1841ed3a06
(Unfortunately, because of API changes I had to bump up minimum IDE version to `192.5728.98`.)

**Comment by dkandalov** [comment]
It was released in plugin version 0.3.8 
https://github.com/dkandalov/limited-wip/releases/tag/v0.3.8
